thunderbird/processes/wps_generate_climos.py

```diff
@@ -0,0 +1,296 @@
+            for line in output:
+                f.write(f"{line}\n")
+
+        return filename
+
+    def collect_climo_files(self):
+        return [file for file in os.listdir(self.workdir) if file.endswith(".nc")]
+
+    def build_meta_link(self, climo_files):
+        meta_link = MetaLink4(
+            "outputs", "Output of netCDF climo files", workdir=self.workdir
+        )
+
+        for file in climo_files:
+            # Create a MetaFile instance, which instantiates a ComplexOutput object.
+            meta_file = MetaFile(f"{file}", "Climatology", fmt=FORMATS.NETCDF)
+            meta_file.file = os.path.join(self.workdir, file)
+            meta_link.append(meta_file)
+
+        return meta_link.xml
+
+    def format_climo(self, climo):
+        if "all" in climo:
+            return self.climos["historical"] + self.climos["futures"]
+
+        # loop over given climo, replace items ('historical', 'futures') with
+        # items from list
+        return list(
+            set(
+                [
+                    item
+                    for c in climo
+                    for item in (self.climos[c] if c in self.climos.keys() else [c])
+                ]
+            )
+        )
+
+    def format_resolutions(self, resolutions):
+        if "all" in resolutions:
+            return [
+                resolution for resolution in self.resolutions if resolution != "all"
+            ]
+
+        return list(set(resolutions))
+
+    def collect_args(self, request):
+        climo = self.format_climo([climo.data for climo in request.inputs["climo"]])
+        operation = request.inputs["operation"][0].data
+        resolutions = self.format_resolutions(
+            [resolution.data for resolution in request.inputs["resolutions"]]
+        )
+        convert_longitudes = request.inputs["convert_longitudes"][0].data
+        split_vars = request.inputs["split_vars"][0].data
+        split_intervals = request.inputs["split_intervals"][0].data
+        dry_run = request.inputs["dry_run"][0].data
+        return (
+            climo,
+            resolutions,
+            convert_longitudes,
+            split_vars,
+            split_intervals,
+            dry_run,
+            operation,
+        )
+
+    def get_filepath(self, request):
+        if "opendap" in request.inputs:
+            return request.inputs["opendap"][0].url
+        elif "netcdf" in request.inputs:
+            return request.inputs["netcdf"][0].file
+        else:
+            raise ProcessError("You must provide a data source")
+
+    def _handler(self, request, response):
+        progress = 0
+        response.update_status("Starting Process", progress)
+
+        (
+            climo,
+            resolutions,
+            convert_longitudes,
+            split_vars,
+            split_intervals,
+            dry_run,
+            operation,
+        ) = self.collect_args(request)
+        filepath = self.get_filepath(request)
+
+        progress += 5
```

Comment:
Do you have any better ways to measure progress here? This seems... dubious. Also, I'm pretty sure that not each independent path through this function will add up to 100.

Reply:
I agree with you, it is not the best way to go about it.  I tried to write a class to make it cleaner but really it didn't help (if anything it made it worse).  Do you have any suggestions on how to improve this?

> Also, I'm pretty sure that not each independent path through this function will add up to 100

As long as the last line has the value `100` the process will be completed (although if you add up progress to `>= 100` it will send the "completed" message early, so still not great).
```python
response.update_status("Process Complete", 100)
```